src/order/countable_dense_linear_order.lean

```diff
@@ -0,0 +1,224 @@
+subtype.map (finset.image (equiv.prod_comm _ _)) $ λ f hf p q hp hq,
+  eq.symm $ hf ((equiv.prod_comm α β).symm p) ((equiv.prod_comm α β).symm q)
+(by { rw [←finset.mem_coe, finset.coe_image, equiv.image_eq_preimage] at hp,
+  rwa ←finset.mem_coe })
+(by { rw [←finset.mem_coe, finset.coe_image, equiv.image_eq_preimage] at hq,
+  rwa ←finset.mem_coe })
+
+/-- The set of partial isomorphisms defined on `b : β `, together with a proof that any
+    partial isomorphism can be extended to include `b`. We prove this by symmetry. -/
+def defined_at_right [densely_ordered α] [no_bot_order α] [no_top_order α] [nonempty α]
+(b : β) : cofinal (partial_iso α β) :=
+{ carrier := λ f, ∃ a, (a, b) ∈ f.val,
+  mem_gt :=
+  begin
+    intro f,
+    rcases (defined_at_left b).mem_gt (partial_iso.comm f) with ⟨f', ⟨a, ha⟩, hl⟩,
+    use partial_iso.comm f',
+    split,
+    { use a,
+      change (a, b) ∈ f'.val.image _,
+      rw [←finset.mem_coe, finset.coe_image, equiv.image_eq_preimage],
+      exact ha },
+    { change _ ⊆ f'.val.image _,
+      rw [←finset.coe_subset, finset.coe_image, equiv.subset_image],
+      change f.val.image _ ⊆ _ at hl,
+      rwa [←finset.coe_subset, finset.coe_image] at hl }
+  end }
+
+end partial_iso
+open partial_iso
+
+variables (α β)
+
+/-- Any countable linear order embeds in any nonempty dense linear order without endpoints. -/
+def embedding_from_countable_to_dense
+[encodable α] [densely_ordered β] [no_bot_order β] [no_top_order β] [nonempty β] :
+  α ↪o β :=
+let our_ideal : ideal (partial_iso α β) :=
+  ideal_of_cofinals (default _) defined_at_left in
+have exists_right : ∀ (a : α), ∃ (b : β) (f ∈ our_ideal), (a, b) ∈ subtype.val f :=
+  begin
+    intro a,
+    rcases cofinal_meets_ideal_of_cofinals (default _) defined_at_left a
+      with ⟨f, ⟨b, hb⟩, hf⟩,
+    exact ⟨b, f, hf, hb⟩,
+  end,
+order_embedding.of_strict_mono (λ a, classical.some (exists_right a))
+begin
+  intros a b,
+  refine (lt_iff_lt_of_cmp_eq_cmp _).mp,
+  rcases classical.some_spec (exists_right a) with ⟨f, hf, ha⟩,
+  rcases classical.some_spec (exists_right b) with ⟨g, hg, hb⟩,
+  rcases our_ideal.directed _ hf _ hg
+    with ⟨m, hm, fm, gm⟩,
+  exact m.property (a, _) (b, _) (fm ha) (gm hb)
+end
+
+/-- Any two countable dense, nonempty linear orders without endpoints are order isomorphic. --/
+def iso_of_countable_dense
+[encodable α] [densely_ordered α] [no_bot_order α] [no_top_order α] [nonempty α]
+[encodable β] [densely_ordered β] [no_bot_order β] [no_top_order β] [nonempty β]
+  : α ≃o β :=
+let to_cofinal : α ⊕ β → cofinal (partial_iso α β) :=
+  λ p, sum.rec_on p defined_at_left defined_at_right in
+let our_ideal : ideal (partial_iso α β) :=
+  ideal_of_cofinals (default _) to_cofinal in
+have exists_right : ∀ (a : α), ∃ (b : β) (f ∈ our_ideal), (a, b) ∈ subtype.val f :=
+  begin
+    intro a,
+    rcases cofinal_meets_ideal_of_cofinals (default _) to_cofinal (sum.inl a)
+      with ⟨f, ⟨b, hb⟩, hf⟩,
+    exact ⟨b, f, hf, hb⟩,
+  end,
```

Comment:
I see a couple of `have exists_right` that are structured almost the same, is it worth it to extract a lemma?

Reply:
I didn't think this would be possible, but I think I found a nice way to do it: see `fun_of_ideal`.

Reply:
Looks good, thanks!